src/python/pants/backend/experimental/jvm/goals/coursier.py

```diff
@@ -0,0 +1,151 @@
+    """
+    resolved_lockfile = await Get(
+        CoursierResolvedLockfile,
+        MavenRequirements,
+        MavenRequirements.create_from_maven_coordinates_fields(
+            fields=(request.target[MavenRequirementsField],),
+        ),
+    )
+    resolved_lockfile_json: bytes = resolved_lockfile.to_json()
+
+    lockfile_sources = await Get(
+        SourceFiles,
+        SourceFilesRequest(
+            [request.target.get(Sources)],
+            for_sources_types=[JvmLockfileSources],
+            enable_codegen=False,
+        ),
+    )
+    if not lockfile_sources.files:
+        # The user defined the target and resolved it, but hasn't created a lockfile yet.
+        # For convenience, create the initial lockfile for them in the default path.
+        return CoursierGenerateLockfileResult(
+            digest=await Get(
+                Digest,
+                CreateDigest(
+                    (
+                        FileContent(
+                            path=os.path.join(
+                                request.target.address.spec_path, "coursier_resolve.lockfile"
+                            ),
+                            content=resolved_lockfile_json,
+                        ),
+                    )
+                ),
+            )
+        )
+
+    # We know that there is at least one lockfile source, and also at most 1 because of
+    # JvmLockfileSources.expected_num_files, so we can blindly grab its only source.
+    source_lockfile_digest_contents = await Get(
+        DigestContents, Digest, lockfile_sources.snapshot.digest
+    )
+    source_lockfile_content = source_lockfile_digest_contents[0]
+    if resolved_lockfile_json != source_lockfile_content.content:
+        # The generated lockfile differs from the existing one, so return the digest of the generated one.
+        return CoursierGenerateLockfileResult(
+            digest=await Get(
+                Digest,
+                CreateDigest(
+                    (
+                        FileContent(
+                            path=source_lockfile_content.path, content=resolved_lockfile_json
+                        ),
+                    )
+                ),
+            )
+        )
+    # The generated lockfile didn't change, so return an empty digest.
+    return CoursierGenerateLockfileResult(
+        digest=EMPTY_DIGEST,
+    )
+
+
+@goal_rule
+async def coursier_resolve_lockfiles(
+    console: Console,
+    targets: Targets,
+    resolve_subsystem: CoursierResolveSubsystem,
+    workspace: Workspace,
+) -> CoursierResolve:
+    jvm_lockfile_targets = Targets(
+        target for target in targets if target.has_field(JvmLockfileSources)
+    )
+    results = await MultiGet(
+        Get(CoursierGenerateLockfileResult, CoursierGenerateLockfileRequest(target=target))
+        for target in jvm_lockfile_targets
+    )
+    results_to_write = tuple(result for result in results if result.digest != EMPTY_DIGEST)
+    if results_to_write:
```

Comment:
Excellent. Not sure if you picked this up by looking at `fmt.py`, but indeed it's important perf wise to not call `workspace.write_digest` if there were no changes. It may be worth a brief comment explaining we only write to the workspace if necessary to avoid a performance hit.

Reply:
I'd gotten it from `fmt.py`, but I actually didn't realize it was for performance, I was doing it avoid unnecessarily churning the mtime of files that didn't need to change!

Added a comment

Reply:
True, that too! Before this optimization, ./pants fmt used to almost always cause IntelliJ to complain that the files changed on disk, along with bad performance to rewrite every file.